### PR TITLE
feat: match metadata

### PR DESCRIPTION
**Pull request by liangyiliang**
# Description

Related issue/PR: #785 

We expose some metadata about a specific _substitution_ (or matching) to the interior of a style block. In particular, two metadata are available:
* `match_id` is the 1-indexed ordinal of the specific matching, and
* `match_total` is the total number of matches.

For example:
```
-- dsl
type MyType

-- sub
MyType T1, T2, T3, T4

-- sty, canvas specifications omitted
forall MyType t {
  // can use `match_id` and `match_total` here.
}
```
This _Style_ block is matched four times, so `match_total` equals to `4` within this style block. Furthermore, `match_id` can be `1`, `2`, `3`, or `4` depending on which matching we are processing.

# Implementation strategy and design decisions

The current algorithm to process a Style block is:
```
for i = 1 .. N {
  processBlock(subst[i], blockStatements)
}
```
where `N` is the total number of substitutions, and `blockStatements` is the list of statements within this Style block.

The new algorithm is:
```
for i = 1 .. N {
  blockStatements.addInFront("match_id = " + i)
  blockStatements.addInFront("match_total = " + N)
  processBlock(subst[i], blockStatements)
}
```

Essentially, we add two fake AST nodes at the beginning of the block statements, assigning the metadata variables to their correct values, depending on the substitution.

# Examples with steps to reproduce them

See: [example](https://3dd0308b.penrose-72l.pages.dev/try/?gist=e07d1ec3de4be8b5577fc1b568479ad8)

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated changes to the `diagrams/` folder

### CODE REVIEW

File: packages/core/src/compiler/Style.ts [hunk 1]
```diff
@@ -1790,15 +1791,32 @@ const resolveLhsName = (
...
 ): Result<ResolvedPath<C>, StyleError> => {
   const { start, end, name, members, indices } = path;
-  return indices.length > 0
-    ? err({ tag: "AssignAccessError", path })
-    : ok({ start, end, ...resolveLhsName(block, assignment, name), members });
+  if (indices.length > 0) {
+    return err({ tag: "AssignAccessError", path });
+  } else {
+    const resolvedLhsName = resolveLhsName(block, assignment, name);
+    /*if (reservedVariableNames.includes(toString(name))) {
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:1806]: Is this useful?

Comment (wodeni) [reply packages/core/src/compiler/Style.ts:1806]: If not, you can revert to the ternary right?

File: packages/core/src/compiler/Style.ts [hunk 2]
```diff
@@ -1872,6 +1890,45 @@ const blockId = (
   }
 };
 
+const makeFakeIntPathAssign = (name: string, value: number): PathAssign<C> => {
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:1893]: TODO: need an API for synthetic node construction

File: packages/core/src/types/errors.ts [hunk 3]
```diff
@@ -134,6 +134,11 @@ export interface DeconstructNonconstructor {
   deconstructor: Deconstructor<A>;
 }
 
+export interface ReadonlyVariableMutationError {
+  tag: "ReadonlyVariableMutationError";
```
Comment (wodeni) [inline packages/core/src/types/errors.ts:138]: Is this in use?

**Review by wodeni — COMMENTED** [review]
Some minor comments below

File: packages/core/src/compiler/Style.ts (hunk 1 above)
Comment (liangyiliang) [reply packages/core/src/compiler/Style.ts]: Yup I can do that

File: packages/core/src/types/errors.ts (hunk 3 above)
Comment (liangyiliang) [reply packages/core/src/types/errors.ts]: Removing that

**Pull request merged**